### PR TITLE
Add (unused) schema parsing based on apollo-compiler

### PR DESCRIPTION
**Pull request by SimonSapin**
Add a `Schema::parse_with_hir` constructor, which should behave the same as the existing `Schema::parse` but is based on apollo-compiler’s HIR instead of apollo-parser’s AST. This new code is not yet used during normal Router operation. Instead it is (for now) only reachable through a semi-public API that runs both code paths in order to compare outputs.

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

Manual testing: https://github.com/apollographql/harness-vs-apollo-rs/pull/1 runs the comparison on a corpus of many schemas used in production. With https://github.com/apollographql/router/pull/2396 also applied, no output differences were found.

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/spec/schema.rs [hunk 1]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
     root_operations: HashMap<OperationKind, String>,
 }
 
+/// YAML-like representation with sorted hashmap/sets, more amenable to diffing
+impl std::fmt::Debug for Schema {
```
Comment (SimonSapin) [inline apollo-router/src/spec/schema.rs:46]: An alternative could be to replace all `HashMap`/`HashSet`s with `IndexMap`/`IndexSet` so that `derive(Debug)` output would be stable, but this might impact runtime performance: https://github.com/bluss/indexmap#performance

File: apollo-router/src/lib.rs [hunk 2]
```diff
@@ -101,4 +101,17 @@ pub mod _private {
     // For tests
     pub use crate::plugins::telemetry::Telemetry as TelemetryPlugin;
     pub use crate::router_factory::create_test_service_factory_from_yaml;
+
+    /// Retuns the `Debug` fomatting of two `Result<Schema, SchemaError>`,
+    /// from `parse_with_ast` and `parse_with_hir` respectively.
+    ///
+    /// The two strings are expected to be equal.
+    pub fn compare_schema_parsing(schema: &str) -> (String, String) {
```
Comment (bnjjj) [inline apollo-router/src/lib.rs:109]: NIT: it's weird to call this function `compare_...` as it doesn't compare anything, it just parses the schema with 2 different ways

File: apollo-router/src/spec/schema.rs [hunk 3]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
...
+                            interfaces: def
+                                .implements_interfaces()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.implements_interfaces()),
+                                )
+                                .map(implements_interface)
+                                .collect(),
```
Comment (bnjjj) [inline apollo-router/src/spec/schema.rs:250]: I'm wondering if it would be more readable if we used `From` impl for that kind of code ?

File: apollo-router/src/spec/schema.rs [hunk 4]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
...
```
Comment (bnjjj) [inline apollo-router/src/spec/schema.rs:284]: I'm wondering if it would be more readable if we used `From` impl for that kind of code ?

File: apollo-router/src/spec/schema.rs [hunk 5]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
...
+                                .input_fields_definition()
+                                .iter()
+                                .chain(
+                                    def.extensions()
+                                        .iter()
+                                        .flat_map(|ext| ext.input_fields_definition()),
+                                )
+                                .map(input_field)
+                                .collect(),
+                        },
```
Comment (bnjjj) [inline apollo-router/src/spec/schema.rs:318]: I'm wondering if it would be more readable if we used `From` impl for that kind of code ?

File: apollo-router/src/spec/schema.rs [hunk 6]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
...
+                })
+                .collect();
+
+            let root_operations = compiler
+                .db
+                .schema()
+                .root_operation_type_definition()
+                .iter()
+                .filter(|def| def.loc().is_some()) // exclude implict operations
+                .map(|def| {
```
Comment (bnjjj) [inline apollo-router/src/spec/schema.rs:342]: `filter_map`

File: apollo-router/src/spec/schema.rs [hunk 7]
```diff
@@ -38,24 +42,353 @@ pub(crate) struct Schema {
...
+                    )
+                })
+                .collect();
+
+            let custom_scalars = compiler
+                .db
+                .scalars()
+                .iter()
+                .filter(|(_name, def)| !def.is_built_in())
+                .map(|(name, _def)| name.clone())
```
Comment (bnjjj) [inline apollo-router/src/spec/schema.rs:362]: `filter_map`

**Comment by bnjjj** [comment]
Please @SimonSapin Could you not override commit and force push to help me re review in a more comfortable way after ?

File: apollo-router/src/lib.rs (hunk 2 above)
Comment (SimonSapin) [reply apollo-router/src/lib.rs:109]: Yeah that’s fair. It returns two strings for the caller to compare. Do you have a suggestion?

That’s left to the caller is so that differences can be printed with https://crates.io/crates/similar-asserts without making that crate a dependency of the router. https://github.com/apollographql/harness-vs-apollo-rs is a separate crate so it needs the router crate to provide some `pub` non-`#[cfg(test)]` API.

Anyway this is a semi-private API that we’ll remove eventually.

File: apollo-router/src/spec/schema.rs (hunk 7 above)
Comment (bnjjj) [reply apollo-router/src/spec/schema.rs:362]: It's just a matter of taste. I prefer this one 
```rust
    .filter_map(|(name, def)| {
        if !def.is_built_in() {
            Some(name.clone())
        } else {
            None
        }
    })
```

but it is very subjective so... as you wish :p

File: apollo-router/src/spec/schema.rs (hunk 3 above)
Comment (SimonSapin) [reply apollo-router/src/spec/schema.rs]: Done! Indeed it looks better

**Review by bnjjj — APPROVED** [review]
LGTM

File: apollo-router/src/query_planner/execution.rs [hunk 8]
```diff
@@ -42,7 +42,7 @@ impl QueryPlan {
         context: &'a Context,
         service_factory: &'a Arc<SubgraphServiceFactory>,
         supergraph_request: &'a Arc<http::Request<Request>>,
-        schema: &'a Schema,
+        schema: &'a Arc<Schema>,
```
Comment (Geal) [inline apollo-router/src/query_planner/execution.rs:45]: Why this change?

File: apollo-router/src/spec/schema.rs [hunk 9]
```diff
@@ -38,24 +42,271 @@ pub(crate) struct Schema {
...
+                        )));
+                    }
+                }
+            }
+
+            let object_types: HashMap<_, _> = compiler
+                .db
+                .object_types()
+                .iter()
+                .map(|(name, def)| (name.clone(), (&**def).into()))
```
Comment (Geal) [inline apollo-router/src/spec/schema.rs:223]: Is there a simpler way to write this?

File: apollo-router/src/query_planner/execution.rs (hunk 8 above)
Comment (SimonSapin) [reply apollo-router/src/query_planner/execution.rs:45]: This enables removing `#[derive(Clone)]` on `Schema`, which in turn is not really necessary for this iteration of this PR since we don’t keep `ApolloCompiler` inside `Schema`. But IMO it’s still better to avoid cloning multiple `HashMap`s and instead clone an `Arc`.

File: apollo-router/src/spec/schema.rs (hunk 9 above)
Comment (SimonSapin) [reply apollo-router/src/spec/schema.rs:223]: I don’t think so, unless the conversion is changed from `impl From<&hir::ObjectTypeDefinition> for ObjectType` to `impl From<&Arc<hir::ObjectTypeDefinition>> for ObjectType` which I’m not a fan of.

File: apollo-router/src/query_planner/execution.rs (hunk 8 above)
Comment (Geal) [reply apollo-router/src/query_planner/execution.rs:45]: oh, right, we only needed a reference before, but with `@defer` the schema is cloned
